data/campaigns/tutorial03_seafaring.wmf/scripting/mission_thread.lua

```diff
@@ -83,24 +100,58 @@ function expedition()
       end
    end
 
-   conclude()
+   conclude_expedition()
 end
 
-function conclude()
+function conclude_expedition()
    additional_port_space.terr = "desert_steppe" -- make it land again so that the player can build a port
-   message_box_objective(plr, conclusion)
-   sleep(5000)
-   waterways()
+   message_box_objective(plr, expedition4)
+   sleep(3000)
+   iron_mine()
+end
+
+function iron_mine()
+   local o = message_box_objective(plr, expedition5)
+   while #plr:get_buildings("atlanteans_ironmine") < 1 do sleep(3000) end
+   set_objective_done(o)
+   expedition_done = true
 end
 
 function complete_ferry_yard()
    plr:allow_buildings{"atlanteans_ferry_yard"}
    local o = message_box_objective(plr, ferry_2)
-   while #plr:get_buildings("atlanteans_ferry_yard") < 1 do sleep(200) end
+   while #plr:get_buildings("atlanteans_ferry_yard") < 1 do sleep(3000) end
    set_objective_done(o)
 end
 
+function complete_gold_mine()
+   local field_for_mine = map:get_field(20, 102):region(5)
+   while field_for_mine do
+      sleep(3000)
+      for i,f in pairs(field_for_mine) do
+         if f.immovable and f.immovable.descr.name == "atlanteans_goldmine" then
+            field_for_mine = nil
+            break
+         end
+      end
+   end
+   gold_mine_done = true
+end
+
+function complete_ferries()
+   -- wait until 4 ferries are assigned to waterways
+   while plr:get_workers("atlanteans_ferry") < 4 do sleep(3000) end
```

Comment:
This only checks that ferries have been produced, not that waterways have been built. This might be good enough though.

Reply:
@gunchleoc 

> This only checks that ferries have been produced, not that waterways have been built.

`plr:get_workers("atlanteans_ferry")` returns only count of ferries *assigned* to waterways. Unemployed ferries are ignored by the function. To be honest, it was a surprise for me.

(Btw, same applies to in-game statistics of total number of your workers.)

That's why I decided to remind player to stop the ferry yard when first ferry is assigned to waterway. This behavior of `get_workers()` function also makes no use of "for-cycle check" code which I therefore removed.

Reply:
Alternative check on produced ferries: [pospec4444:3675-count-ferries](https://github.com/pospec4444/widelands/compare/3675-more-seafaring-tutorial-improvements...pospec4444:3675-count-ferries)

Reply:
@Noordfrees I'd like your opinion on the get_workers function here, since you wrote the ferry code.

Reply:
This is very fragile regarding ferries. Unlike all other worker types they are not managed by worker economies and therefore explicitly do *not* allow you to assert that their `worker_economy_` always points to an existing Economy instance, or that they are added or removed to worker economies correctly. Therefore this *might* work, but it might just as well contain some evil corner-case bugs, so I strongly recommend using a function to count ferries on the map instead.

Reply:
@Noordfrees
Is it safe to use `get_workers("atlanteans_ferry")` to get number of ferries **operating any waterways**?

Reply:
No. I just rechecked `Ferry::set_economy` and the most important calling functions – the ferry's economies are not being maintained at all, neither when idle nor when assigned to a WW, therefore I cannot imagine that economies really know how many ferries exist. It is to be expected that `get_workers("atlanteans_ferry")` may show undefined behaviour.